File: V-REP/programming/remoteApiBindings/lua/lua/complexCommandTest.lua
-- This example illustrates how to execute complex commands from
-- a remote API client. You can also use a similar construct for
-- commands that are not directly supported by the remote API.
--
-- Load the demo scene 'remoteApiCommandServerExample.ttt' in V-REP, then 
-- start the simulation and run this program.
--
-- IMPORTANT: for each successful call to simxStart, there
-- should be a corresponding call to simxFinish at the end!

print('Program started')
require 'remoteApiLua'
simxFinish(-1) -- just in case, close all opened connections
local clientID=simxStart('127.0.0.1',19999,true,true,2000,5)
if clientID~=-1 then
    print('Connected to remote API server')

    -- 1. First send a command to display a specific message in a dialog box:
    local result,intsOut,floatsOut,stringsOut,bufferOut=simxCallScriptFunction(clientID,'remoteApiCommandServer',sim_scripttype_childscript,'displayText_function',{},{},{'Hello world'},'',simx_opmode_blocking)
    if result==simx_return_ok then
        print('Returned message: '..stringsOut[1]) -- display the reply from V-REP (in this case, just a string)
    else
        print('Remote function call failed')
    end

    -- 2. Now create a dummy object at coordinate 0.1,0.2,0.3 with name 'MyDummyName':
    local result,intsOut,floatsOut,stringsOut,bufferOut=simxCallScriptFunction(clientID,'remoteApiCommandServer',sim_scripttype_childscript,'createDummy_function',{},{0.1,0.2,0.3},{'MyDummyName'},'',simx_opmode_blocking)
    if result==simx_return_ok then
        print('Dummy handle: '..intsOut[1]) -- // display the reply from V-REP (in this case, the handle of the created dummy)
    else
        print('Remote function call failed')
    end

    -- 3. Now send a code string to execute some random functions:
    local code=[[
local octreeHandle=simCreateOctree(0.5,0,1)
simInsertVoxelsIntoOctree(octreeHandle,0,{0.1,0.1,0.1},{255,0,255})
return 'done'
]]
    local result,intsOut,floatsOut,stringsOut,bufferOut=simxCallScriptFunction(clientID,'remoteApiCommandServer',sim_scripttype_childscript,'executeCode_function',{},{},{code},'',simx_opmode_blocking)
    if result==simx_return_ok then
        print('Code execution returned: '..stringsOut[1])
    else
        print('Remote function call failed')
    end

    -- Now close the connection to V-REP:   
    simxFinish(clientID)
else
    print('Failed connecting to remote API server')
end
print('Program ended')

File: V-REP/programming/remoteApiBindings/java/java/complexCommandTest.java
// This example illustrates how to execute complex commands from
// a remote API client. You can also use a similar construct for
// commands that are not directly supported by the remote API.
//
// Load the demo scene 'remoteApiCommandServerExample.ttt' in V-REP, then 
// start the simulation and run this program.
//
// IMPORTANT: for each successful call to simxStart, there
// should be a corresponding call to simxFinish at the end!

import coppelia.IntWA;
import coppelia.FloatWA;
import coppelia.StringWA;
import coppelia.remoteApi;

public class complexCommandTest
{
    public static void main(String[] args)
    {
        System.out.println("Program started");
        remoteApi vrep = new remoteApi();
        vrep.simxFinish(-1); // just in case, close all opened connections
        int clientID = vrep.simxStart("127.0.0.1",19999,true,true,5000,5);
        if (clientID!=-1)
        {
            System.out.println("Connected to remote API server");   

            // 1. First send a command to display a specific message in a dialog box:
            StringWA inStrings=new StringWA(1);
            inStrings.getArray()[0]="Hello world!";
            StringWA outStrings=new StringWA(0);
            int result=vrep.simxCallScriptFunction(clientID,"remoteApiCommandServer",vrep.sim_scripttype_childscript,"displayText_function",null,null,inStrings,null,null,null,outStrings,null,vrep.simx_opmode_blocking);
            if (result==vrep.simx_return_ok)
                System.out.format("Returned message: %s\n",outStrings.getArray()[0]); // display the reply from V-REP (in this case, just a string)
            else
                System.out.format("Remote function call failed\n");

            // 2. Now create a dummy object at coordinate 0.1,0.2,0.3 with name 'MyDummyName':
            FloatWA inFloats=new FloatWA(3);
            inFloats.getArray()[0]=0.1f;
            inFloats.getArray()[1]=0.2f;
            inFloats.getArray()[2]=0.3f;
            inStrings=new StringWA(1);
            inStrings.getArray()[0]="MyDummyName";
            IntWA outInts=new IntWA(0);
            result=vrep.simxCallScriptFunction(clientID,"remoteApiCommandServer",vrep.sim_scripttype_childscript,"createDummy_function",null,inFloats,inStrings,null,outInts,null,null,null,vrep.simx_opmode_blocking);
            if (result==vrep.simx_return_ok)
                System.out.format("Dummy handle: %d\n",outInts.getArray()[0]); // display the reply from V-REP (in this case, the handle of the created dummy)
            else
                System.out.format("Remote function call failed\n");

            // 3. Now send a code string to execute some random functions:
            inStrings.getArray()[0]="local octreeHandle=simCreateOctree(0.5,0,1)\n" 
            + "simInsertVoxelsIntoOctree(octreeHandle,0,{0.1,0.1,0.1},{255,0,255})\n" 
            + "return 'done'";
            result=vrep.simxCallScriptFunction(clientID,"remoteApiCommandServer",vrep.sim_scripttype_childscript,"executeCode_function",null,null,inStrings,null,null,null,outStrings,null,vrep.simx_opmode_blocking);
            if (result==vrep.simx_return_ok)
                System.out.format("Code execution returned: %s\n",outStrings.getArray()[0]);
            else
                System.out.format("Remote function call failed\n");

            // Now close the connection to V-REP:   
            vrep.simxFinish(clientID);
        }
        else
            System.out.println("Failed connecting to remote API server");
        System.out.println("Program ended");
    }
}
